Bareborns registry for encoders

jankounchained opened the pull request:
Not to be merged in yet.

a) Is this the right direction with the encoder registry?
Can we use turtopic classes, or should there be a topic-benchmark base class for encoders?
PS [E5-KeyNMF encoder](https://github.com/x-tabdeveloping/turftopic/issues/23) will be added later.


b) The `encoder_registry` is empty.
Any idea WTF is going on?

Trying:
```
from topic_benchmark.registries import encoder_registry
print(encoder_registry.get_all())
```
Returns `{}`

jankounchained committed:
Bareborns registry for encoders

fix: warn user about defaulting to ST

x-tabdeveloping commented:
a) Turftopic classes are cool! It's looking really good!
b) I'm not sure why it happens, but I had similar issues, and it's something I encounter on a regular basis with Catalogue. Does it happen in the CLI as well? My guess is that it could actually work in the CLI code cause it somehow gets imported. I will take a look for you though.

jankounchained commented:
thanks! yeah it happens in the CLI as well..

maybe the "solution is to use entry points"
https://github.com/explosion/catalogue?tab=readme-ov-file#how-do-i-make-sure-all-of-the-registration-decorators-have-run
tried to understand it yesterday, ain't got it yet

i also see that kenneth has a function in seb for "making custom registered functions available"
https://github.com/KennethEnevoldsen/scandinavian-embedding-benchmark/blob/main/src/seb/cli/import_code.py
i'll ask him

x-tabdeveloping commented on topic_benchmark/cli.py:
```diff
@@ -22,11 +24,23 @@ def run_cli(
+    # if not found, assume encoder is a sentence transformer
+    try:
+        encoder = encoder_registry.get(encoder_model)
+    except:
```
Bare except is not a good idea. There is info on why [here](https://stackoverflow.com/questions/54948548/what-is-wrong-with-using-a-bare-except). (For instance it would catch a `KeyboardInterrupt`, which we don't want)
Also I think this wouldn't fail here, because `.get()` might just return `None` if it is not present as far as I know.
The more specific we are with the `except` clause the better. In this case my guess is that it's a `KeyError`.

x-tabdeveloping commented on topic_benchmark/cli.py:
```diff
@@ -22,11 +24,23 @@ def run_cli(
+        encoder = encoder_registry.get(encoder_model)
+    except:
+        encoder = SentenceTransformer(encoder_model)
+        warnings.warn(
```
Does this need to be a warning? Isn't this more an informative message, than something the user should be concerned about?

x-tabdeveloping committed:
Added encoders to __init__, registry will probably work now.

x-tabdeveloping committed:
Added RegistryError instead of bare except

Changed try-catch to guard if statement

fix: properly initalize encoder class upon retrieval from registry, change warning to simple message

x-tabdeveloping commented:
I think I managed to fix the registry error, can you confirm?

jankounchained commented:
added one small thing, now it works. 
rebased, ready for merging

x-tabdeveloping commented:
Awesome, feel free to merge!
